### DISCUSSION TITLE
Css styling changes on refresh only in vercel deployment

### DISCUSSION THREAD

**Discussion by c2bsolutions** [vercel/next.js#39804] [Help] [answered]
Has anyone had an issue with css or styling changing on refresh when deployed to vercel? 

This only happens on refresh in my vercel deployment and I cannot recreate the issue in my local environment.

The main components that changes are in the web3uikit.

Thanks in Advance!

**Comment by icyJoseph** [comment, marked as answer]
Hi,

Have you got a link to your project?

What is exactly happening? The styles are missing when you refresh? how are they `changing`? I wonder if your styles are being calculated at runtime. It'd be odd that it works locally though. Does it work locally with `next build && next start`?

## Edit

OP's repo is missing the `_document` configuration required when using Styled Components, in order to calculate the styles are build time. 

- https://github.com/vercel/next.js/blob/canary/examples/with-styled-components/pages/_document.tsx
- https://github.com/vercel/next.js/blob/canary/examples/with-styled-components-babel/pages/_document.tsx
- https://github.com/chemicalkosek/next.js/blob/467758b9fcebfe88e68cdf4b97c815a5c1c16936/examples/with-styled-components/pages/_document.js (JavaScript example)

**Reply by c2bsolutions** [reply]
Yes my local build is working with next build but the form structure of the page changes when I refresh the page at https://funds-for-humanity-polygon-testnet.vercel.app/CreateFund 

I noticed that it only happens for unpopulated fields with the page loads.

Thanks for you feedback!

**Reply by icyJoseph** [reply]
Let's say I visit  your page with JS disabled, then I see this.

![Screenshot 2022-08-22 at 16 57 05](https://user-images.githubusercontent.com/21013447/185952851-6e2e4c67-89bb-4840-923e-5dc296fa4679.png)

What stack are you using to handle CSS? Styled Components, Tailwind? Then you need to use - https://github.com/vercel/next.js/blob/canary/examples/with-styled-components/pages/_document.tsx, this modification to the _document element to pre-compile the styles and ship them as part of the initial HTML. This might not be the issue though, but good to check that it is in place. 

`next.config.js` also needs to have its compiler options set to styled-components, https://nextjs.org/docs/advanced-features/compiler#styled-components.

Otherwise, it seems to be that they compile client side, and that leads to this flash of unstyled components. Still looks fishy, this is not how it should look like. 

Do you have a link to a repository to take a further look?

**Reply by c2bsolutions** [reply]
Yes thank you, I am using tailwind for my css. This has been very helpful and I will take a dive into these suggestions! You can access my repository at https://github.com/c2bsolutions/FundsForHumanityPolygonTestnetFrontEnd

**Reply by icyJoseph** [reply]
Hi,

As I suspected, the problem is:

> Usage with Next.js
> If you are using web3uikit with Next.js, be sure to follow the [official guide](https://github.com/vercel/next.js/tree/canary/examples/with-styled-components-babel), since we are using styled-components under the hood.

Which means that **you are missing `_document` under pages**. And also your `next.config.js`, should implement the `styledComponents` compiler, **which you are already have**.

See how they do it here: https://github.com/vercel/next.js/blob/canary/examples/with-styled-components-babel/pages/_document.tsx,

**Reply by c2bsolutions** [reply]
ok thank you, does this mean that web3uikit only render correctly in typescript when deploying to vercel?

**Reply by icyJoseph** [reply]
Ah, no, do not get confused, TypeScript/JavaScript doesn't really matter here.

What has to happen is that Styled Components (used by web3uikit), needs to walk your application, at build time, to collect the stylesheets used, so that these can be shipped with the first UI load.

You can make a `_document.js` file, and do the same there: https://github.com/chemicalkosek/next.js/blob/467758b9fcebfe88e68cdf4b97c815a5c1c16936/examples/with-styled-components/pages/_document.js (a JavaScript file)

Again, this has nothing to do with TypeScript or Javascript. It is part of understanding how SSR, and CSS-in-JS solutions like Styled Components work within that world.

**Reply by c2bsolutions** [reply]
That resolved my issue. Thank you very much icyJoseph!

**Reply by icyJoseph** [reply]
No worries! Please do not forget to mark as answered! I updated my initial comment with relevant info.
